# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsFrameLoaderOwner.cpp
# Commit: 30d2bf04719b
# Full Hash: 30d2bf04719bfb60095aaa7d679363e57d25ad92
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-06-13 03:21:52
# Regressor Bug: 1640019
# File Overlap Count: 1
# Description:
#   Bug 1640019 - Part 1: Support toplevel process switches outside of tabbrowser, r=mattwoodrow
#   
#   This new process switching behavior is only enabled for some browser elements,
#   which have specified a specific attribute. Turning this on for all browsers with
#   a `remote` attribute causes breakage in reftests.
# ==============================================================================

diff -r 3fefa1d023a6 -r 30d2bf04719b dom/base/nsFrameLoaderOwner.cpp
--- a/dom/base/nsFrameLoaderOwner.cpp	Fri Jun 12 17:02:45 2020 +0000
+++ b/dom/base/nsFrameLoaderOwner.cpp	Fri Jun 12 16:52:00 2020 +0000
@@ -49,33 +49,37 @@
   return loadContext->UseRemoteSubframes();
 }
 
-bool nsFrameLoaderOwner::ShouldPreserveBrowsingContext(
-    const mozilla::dom::RemotenessOptions& aOptions) {
-  if (aOptions.mReplaceBrowsingContext) {
-    return false;
+nsFrameLoaderOwner::ChangeRemotenessContextType
+nsFrameLoaderOwner::ShouldPreserveBrowsingContext(
+    bool aIsRemote, bool aReplaceBrowsingContext) {
+  if (aReplaceBrowsingContext) {
+    return ChangeRemotenessContextType::DONT_PRESERVE_BUT_PROPAGATE;
   }
 
   if (XRE_IsParentProcess()) {
     // Don't preserve for remote => parent loads.
-    if (aOptions.mRemoteType.IsVoid()) {
-      return false;
+    if (!aIsRemote) {
+      return ChangeRemotenessContextType::DONT_PRESERVE;
     }
 
     // Don't preserve for parent => remote loads.
     if (mFrameLoader && !mFrameLoader->IsRemoteFrame()) {
-      return false;
+      return ChangeRemotenessContextType::DONT_PRESERVE;
     }
   }
 
   // We will preserve our browsing context if either fission is enabled, or the
   // `preserve_browsing_contexts` pref is active.
-  return UseRemoteSubframes() ||
-         StaticPrefs::fission_preserve_browsing_contexts();
+  if (UseRemoteSubframes() ||
+      StaticPrefs::fission_preserve_browsing_contexts()) {
+    return ChangeRemotenessContextType::PRESERVE;
+  }
+  return ChangeRemotenessContextType::DONT_PRESERVE;
 }
 
 void nsFrameLoaderOwner::ChangeRemotenessCommon(
     const ChangeRemotenessContextType& aContextType,
-    bool aSwitchingInProgressLoad, const nsAString& aRemoteType,
+    bool aSwitchingInProgressLoad, bool aIsRemote,
     std::function<void()>& aFrameLoaderInit, mozilla::ErrorResult& aRv) {
   RefPtr<mozilla::dom::BrowsingContext> bc;
   bool networkCreated = false;
@@ -120,7 +124,7 @@
     }
 
     mFrameLoader = nsFrameLoader::Recreate(
-        owner, bc, aRemoteType, networkCreated,
+        owner, bc, aIsRemote, networkCreated,
         aContextType == ChangeRemotenessContextType::PRESERVE);
     if (NS_WARN_IF(!mFrameLoader)) {
       aRv.Throw(NS_ERROR_FAILURE);
@@ -179,7 +183,13 @@
 
 void nsFrameLoaderOwner::ChangeRemoteness(
     const mozilla::dom::RemotenessOptions& aOptions, mozilla::ErrorResult& rv) {
+  bool isRemote = !aOptions.mRemoteType.IsEmpty();
+
   std::function<void()> frameLoaderInit = [&] {
+    if (isRemote) {
+      mFrameLoader->ConfigRemoteProcess(aOptions.mRemoteType, nullptr);
+    }
+
     if (aOptions.mPendingSwitchID.WasPassed()) {
       mFrameLoader->ResumeLoad(aOptions.mPendingSwitchID.Value());
     } else {
@@ -187,16 +197,10 @@
     }
   };
 
-  ChangeRemotenessContextType preserveType =
-      ChangeRemotenessContextType::DONT_PRESERVE;
-  if (ShouldPreserveBrowsingContext(aOptions)) {
-    preserveType = ChangeRemotenessContextType::PRESERVE;
-  } else if (aOptions.mReplaceBrowsingContext) {
-    preserveType = ChangeRemotenessContextType::DONT_PRESERVE_BUT_PROPAGATE;
-  }
-
-  ChangeRemotenessCommon(preserveType, aOptions.mSwitchingInProgressLoad,
-                         aOptions.mRemoteType, frameLoaderInit, rv);
+  auto shouldPreserve =
+      ShouldPreserveBrowsingContext(isRemote, aOptions.mReplaceBrowsingContext);
+  ChangeRemotenessCommon(shouldPreserve, aOptions.mSwitchingInProgressLoad,
+                         isRemote, frameLoaderInit, rv);
 }
 
 void nsFrameLoaderOwner::ChangeRemotenessWithBridge(BrowserBridgeChild* aBridge,
@@ -214,12 +218,26 @@
     mFrameLoader->mRemoteBrowser = host;
   };
 
-  // NOTE: We always use the DEFAULT_REMOTE_TYPE here, because we don't actually
-  // know the real remote type, and don't need to, as we're a content process.
-  ChangeRemotenessCommon(
-      /* preserve */ ChangeRemotenessContextType::PRESERVE,
-      /* switching in progress load */ true,
-      NS_LITERAL_STRING(DEFAULT_REMOTE_TYPE), frameLoaderInit, rv);
+  ChangeRemotenessCommon(ChangeRemotenessContextType::PRESERVE,
+                         /* inProgress */ true,
+                         /* isRemote */ true, frameLoaderInit, rv);
+}
+
+void nsFrameLoaderOwner::ChangeRemotenessToProcess(
+    ContentParent* aContentParent, uint64_t aPendingSwitchId,
+    bool aReplaceBrowsingContext, mozilla::ErrorResult& rv) {
+  MOZ_ASSERT(XRE_IsParentProcess());
+
+  std::function<void()> frameLoaderInit = [&] {
+    mFrameLoader->ConfigRemoteProcess(aContentParent->GetRemoteType(),
+                                      aContentParent);
+    mFrameLoader->ResumeLoad(aPendingSwitchId);
+  };
+
+  auto shouldPreserve = ShouldPreserveBrowsingContext(
+      /* isRemote */ true, aReplaceBrowsingContext);
+  ChangeRemotenessCommon(shouldPreserve, /* inProgress */ true,
+                         /* isRemote */ true, frameLoaderInit, rv);
 }
 
 void nsFrameLoaderOwner::SubframeCrashed() {
@@ -246,6 +264,7 @@
         }));
   };
 
-  ChangeRemotenessCommon(ChangeRemotenessContextType::PRESERVE, false,
-                         VoidString(), frameLoaderInit, IgnoreErrors());
+  ChangeRemotenessCommon(ChangeRemotenessContextType::PRESERVE,
+                         /* inProgress */ false,
+                         /* isRemote */ false, frameLoaderInit, IgnoreErrors());
 }